File: ios/Classes/L2CapChannelManager.swift
// Copyright 2023, Continental Automotive Technologies GmbH
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import Foundation

#if os(macOS)
import FlutterMacOS
#else
import Flutter
#endif


@objc
private enum LogLevel: Int {
    case none = 0
    case error = 1
    case warning = 2
    case info = 3
    case debug = 4
    case verbose = 5
}

@objc
private class LogUtil: NSObject {

    @objc
    public static var LOG_LEVEL: LogLevel = LogLevel.debug

    @objc
    public static func log(logLevel: LogLevel, message: String) {
        if (logLevel.rawValue <= LOG_LEVEL.rawValue) {
            NSLog("[FBP] %@", message)
        }
    }
}

@available(iOS 13.0, *)
@objc
public class L2CapChannelManager : NSObject, CBPeripheralManagerDelegate {

    private var peripheralManager: CBPeripheralManager?
    private var openL2CapChannelInfos: [L2CapServerInfo] = []
    
    private var openL2CapChannelCallback: FlutterResult?
    private var listenL2CapChannelCallback: FlutterResult?
    private var closeL2CapChannelCallback: FlutterResult?
    
    private let deviceConnectedMethodChannel: FlutterMethodChannel
    
    private var peripheralManagerStateContinuation: CheckedContinuation<CBManagerState, Error>?
    
    @objc
    public init(deviceConnectedMethodChannel: FlutterMethodChannel) {
        self.deviceConnectedMethodChannel = deviceConnectedMethodChannel
    }
    
    @objc
    public func listenUsingL2capChannel(request: ListenL2CapChannelRequest, resultCallback: @escaping FlutterResult) {
        Task {
            do {
                let state = try await getPeripheralState()
                if state != CBManagerState.poweredOn {
                    resultCallback(FlutterError(code: ErrorCodes.bluetoothTurnedOff, message: "peripheralManager is not in poweredOn state.", details: nil))
                    return
                }
                listenL2CapChannelCallback = resultCallback
                requirePeripheralManager().publishL2CAPChannel(withEncryption: request.secure)
            } catch {
                resultCallback(
                    FlutterError(
                        code: ErrorCodes.bluetoothTurnedOff,
                        message: error.localizedDescription,
                        details: nil
                    )
                )
            }
        
        }
        
    }
    
    @objc
    public func connectToL2CapChannel(device: CBPeripheral, request: OpenL2CapChannelRequest, resultCallback: @escaping FlutterResult) {
        guard UUID(uuidString: request.remoteId) != nil else {
          LogUtil.log(logLevel: LogLevel.debug, message: String(format: "Provided device identifier is no UUID: %s", request.remoteId))
          resultCallback(FlutterError(code: ErrorCodes.openL2CapChannelFailed, message: "Provided device identifier is not a valid UUID", details: nil))
          return
        }
        
        openL2CapChannelCallback = resultCallback
        device.openL2CAPChannel(CBL2CAPPSM(request.psm))
    }
    
    @objc
    public func read(request: ReadL2CapChannelRequest, resultCallback: FlutterResult) {
        let psm = CBL2CAPPSM(request.psm)
        guard let openChannel = openL2CapChannelInfos.first(where: {$0.getPSM() == psm}) else {
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "No open channel found with PSM: %d", psm))
            resultCallback(FlutterError(code: ErrorCodes.noOpenL2CapChannelFound, message: String(format: "No open channel found with PSM: %d", psm), details: nil))
            return
        }
        guard let deviceUUID = UUID(uuidString: request.remoteId) else {
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "Provided device identifier is no UUID: %s", request.remoteId))
            resultCallback(FlutterError(code: ErrorCodes.inputStreamReadFailed, message: String(format: "Provided device identifier is no UUID: %s", request.remoteId), details: nil))
            return
        }
        openChannel.read(deviceIdentifier: deviceUUID, result: resultCallback)
    }
    
    @objc
    public func write(request: WriteL2CapChannelRequest, resultCallback: FlutterResult) {
        let psm = CBL2CAPPSM(request.psm)
        guard let openChannel = openL2CapChannelInfos.first(where: {$0.getPSM() == psm}) else {
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "No open channel found with PSM: %d", psm))
            resultCallback(FlutterError(code: ErrorCodes.noOpenL2CapChannelFound, message: String(format: "No open channel found with PSM: %d", psm), details: nil))
            return
        }
        guard let deviceUUID = UUID(uuidString: request.remoteId) else {
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "Provided device identifier is no UUID: %s", request.remoteId))
            resultCallback(FlutterError(code: ErrorCodes.outputStreamWriteFailed, message: String(format: "Provided device identifier is no UUID: %s", request.remoteId), details: nil))
            return
        }
        openChannel.write(deviceIdentifier: deviceUUID, payload: request.value, result: resultCallback)
    }
    
    @objc
    public func closeChannel(request: CloseL2CapChannelRequest,  resultCallback: @escaping FlutterResult) {
        let psm = CBL2CAPPSM(request.psm)
        guard let openChannel = openL2CapChannelInfos.first(where: {$0.getPSM() == psm}) else {
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "No open channel found with PSM: %d", psm))
            resultCallback(FlutterError(code: ErrorCodes.noOpenL2CapChannelFound, message: "No open channel found for device and psm", details: nil))
            return
        }
        guard let deviceUUID = UUID(uuidString: request.remoteId) else {
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "Provided device identifier is no UUID: %s", request.remoteId))
            return
        }
        openChannel.close(deviceIdentifier: deviceUUID)
        resultCallback(nil)
    }
    
    @objc
    public func closeServerSocket(request: CloseL2CapServer, resultCallback: @escaping FlutterResult) {
        let peripheralManager = requirePeripheralManager()
        closeL2CapChannelCallback = resultCallback
        let psmToClose = CBL2CAPPSM(request.psm)
        LogUtil.log(logLevel: LogLevel.debug, message: String(format: "Closing L2CapChannel with PSM %d", psmToClose))
        if let channelToCloseIndex = openL2CapChannelInfos.firstIndex(where: { $0.getPSM() == psmToClose}) {
            openL2CapChannelInfos[channelToCloseIndex].closeAllConnections()
            openL2CapChannelInfos.remove(at: channelToCloseIndex)
        }
        peripheralManager.unpublishL2CAPChannel(psmToClose)
    }
    
    private func requirePeripheralManager() -> CBPeripheralManager {
        if peripheralManager == nil {
            peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
        }
        return peripheralManager!
    }
    
    private func getPeripheralState() async throws -> CBManagerState {
        let peripheralManager = requirePeripheralManager()
        if peripheralManager.state == .unknown {
            return try await withCheckedThrowingContinuation { continuation in
                peripheralManagerStateContinuation = continuation
            }
        }
        return peripheralManager.state
    }
 
    // MARK: Peripheral Manager Delegates
    public func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        LogUtil.log(logLevel: LogLevel.debug, message: "peripheralManagerDidUpdateState called.")
        peripheralManagerStateContinuation?.resume(returning: peripheral.state)
        peripheralManagerStateContinuation = nil
    }
    
    public func peripheralManager(_ peripheral: CBPeripheralManager, didPublishL2CAPChannel PSM: CBL2CAPPSM, error: Error?) {
        if let error = error {
            NSLog("Published L2Cap channel failed: %s", error.localizedDescription)
            LogUtil.log(logLevel: LogLevel.error, message: String(format: "Published L2Cap channel failed: %s", error.localizedDescription))
            guard let listenL2CapChannelCallback = listenL2CapChannelCallback else { return }
            listenL2CapChannelCallback(error)
        } else {
            let response = ListenL2CapChannelResponse(psm: Int(PSM))
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "Published L2Cap channel with PSM %d successfully", PSM))
            openL2CapChannelInfos.append(L2CapServerInfo(psm: PSM))
            
            guard let listenL2CapChannelCallback = listenL2CapChannelCallback else { return }
            listenL2CapChannelCallback(response.marshal())
        }
    }
    
    public func peripheralManager(_ peripheral: CBPeripheralManager, didUnpublishL2CAPChannel PSM: CBL2CAPPSM, error: Error?) {
        LogUtil.log(logLevel: LogLevel.debug, message: "ClosedL2Cap channel.")
        guard let closeL2CapChannelCallback = closeL2CapChannelCallback else { return }
        closeL2CapChannelCallback(nil)
    }
    
    public func peripheralManager(_ peripheral: CBPeripheralManager, didOpen channel: CBL2CAPChannel?, error: Error?) {
        LogUtil.log(logLevel: LogLevel.debug, message: "L2Cap Channel opened.")
        if let error = error {
            LogUtil.log(logLevel: LogLevel.error, message: String(format: "didOpenL2CapChannel returns error: %@", error.localizedDescription))
            openL2CapChannelCallback?(FlutterError(code: ErrorCodes.openL2CapChannelFailed, message: error.localizedDescription, details: nil))
        } else {
            guard let channel = channel else {
                LogUtil.log(logLevel: LogLevel.error, message: "No L2Cap channel provided. This should not happen.")
                return
            }
            handleNewConnection(channel: channel)
            openL2CapChannelCallback?(["psm": Int(channel.psm)])
        }
    }
    
    @objc
    public func didOpenChannel(peripheral: CBPeripheral, channel: CBL2CAPChannel?, error: (any Error)?) {
        if let error = error {
            LogUtil.log(logLevel: LogLevel.error, message: String(format: "didOpenL2CapChannel returns error: %@", error.localizedDescription))
            openL2CapChannelCallback?(FlutterError(code: ErrorCodes.openL2CapChannelFailed, message: error.localizedDescription, details: nil))
        } else {
            guard let channel = channel else {
                LogUtil.log(logLevel: LogLevel.error, message: "No L2Cap channel provided. This should not happen.")
                return
            }
            LogUtil.log(logLevel: LogLevel.debug, message: String(format: "L2Cap Channel %d opened.", Int(channel.psm)))
            handleNewConnection(channel: channel)
            openL2CapChannelCallback?(["psm": Int(channel.psm)])
        }
    }
    
    private func handleNewConnection(channel : CBL2CAPChannel) {
        if (!openL2CapChannelInfos.contains(where:  { $0.getPSM() == channel.psm})){
            let newChannelInfo = L2CapServerInfo(psm: channel.psm)
            openL2CapChannelInfos.append(newChannelInfo)
        }
        guard let channelInfo = openL2CapChannelInfos.first(where: { $0.getPSM() == channel.psm}) else {
            return
        }
        channelInfo.addConnection(l2CapChannel: channel)
        
        let event = DeviceConnectedToL2CapChannel(device: channel.peer, psm: Int(channelInfo.getPSM()))
        deviceConnectedMethodChannel.invokeMethod(L2CapMethodNames.deviceConnected, arguments: event.marshal())
    }
    
}

File: android/src/main/java/com/lib/flutter_blue_plus/l2cap/L2CapChannelManager.java
// Copyright 2024, Continental Automotive Technologies GmbH
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
package com.lib.flutter_blue_plus.l2cap;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothServerSocket;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.lib.flutter_blue_plus.ErrorCodes;
import com.lib.flutter_blue_plus.l2cap.channel.L2CapChannel;
import com.lib.flutter_blue_plus.l2cap.channel.L2CapClientChannel;
import com.lib.flutter_blue_plus.l2cap.info.ClientSocketInfo;
import com.lib.flutter_blue_plus.l2cap.info.L2CapInfo;
import com.lib.flutter_blue_plus.l2cap.info.ServerSocketInfo;
import com.lib.flutter_blue_plus.l2cap.messages.CloseL2CapChannelRequest;
import com.lib.flutter_blue_plus.l2cap.messages.CloseL2CapServer;
import com.lib.flutter_blue_plus.l2cap.messages.ListenL2CapChannelRequest;
import com.lib.flutter_blue_plus.l2cap.messages.ListenL2CapChannelResponse;
import com.lib.flutter_blue_plus.l2cap.messages.OpenL2CapChannelRequest;
import com.lib.flutter_blue_plus.l2cap.messages.ReadL2CapChannelRequest;
import com.lib.flutter_blue_plus.l2cap.messages.WriteL2CapChannelRequest;
import com.lib.flutter_blue_plus.log.LogLevel;

import java.io.IOException;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import io.flutter.plugin.common.MethodChannel.Result;

@TargetApi(Build.VERSION_CODES.Q)
public class L2CapChannelManager {
    private final BluetoothAdapter adapter;
    private final DeviceConnected deviceConnectedCallback;
    private final List<L2CapInfo> openL2CapChannelInfos = Collections.synchronizedList(new LinkedList<>());

    public L2CapChannelManager(@NonNull final BluetoothAdapter adapter, @NonNull final DeviceConnected deviceConnectedCallback) {
        this.adapter = adapter;
        this.deviceConnectedCallback = deviceConnectedCallback;
    }

    @SuppressLint("MissingPermission")
    public synchronized void listenUsingL2capChannel(ListenL2CapChannelRequest request, final Result resultCallback) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            firePlatformNotSupportedError(resultCallback);
            return;
        }
        if (!adapter.isEnabled()) {
            LogLevel.DEBUG.log("Bluetooth is disabled. Please enable first.");
            resultCallback.error(ErrorCodes.BLUETOOTH_TURNED_OFF, "Bluetooth is turned off.", null);
            return;
        }

        try {
            final BluetoothServerSocket serverSocket;
            if (request.secure) {
                serverSocket = adapter.listenUsingL2capChannel();
            } else {
                serverSocket = adapter.listenUsingInsecureL2capChannel();
            }
            final ServerSocketInfo socketInfo = new ServerSocketInfo(serverSocket, deviceConnectedCallback);
            openL2CapChannelInfos.add(socketInfo);
            final int psm = serverSocket.getPsm();
            socketInfo.acceptConnections();

            final ListenL2CapChannelResponse response = new ListenL2CapChannelResponse(psm);
            resultCallback.success(response.marshal());

        } catch (IOException e) {
            LogLevel.ERROR.log(e.getMessage(), e);
            resultCallback.error(ErrorCodes.OPEN_L2CAP_CHANNEL_FAILED, e.getMessage(), e);
        }

    }

    public synchronized void connectToL2CapChannel(final OpenL2CapChannelRequest request, final Result resultCallback) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            firePlatformNotSupportedError(resultCallback);
            return;
        }
        final String deviceId = request.remoteId;
        final BluetoothDevice device = adapter.getRemoteDevice(deviceId);
        final int psm = request.psm;
        final boolean secure = request.secure;

        L2CapInfo l2CapInfo = findInfo(psm);
        if (l2CapInfo == null) {
            LogLevel.DEBUG.log("L2CAP Channel with for device " + device.getAddress() + " / psm " + psm + " not open yet. Create channel.");
            l2CapInfo = new ClientSocketInfo(new L2CapClientChannel(device, psm));
            openL2CapChannelInfos.add(l2CapInfo);
        }
        final L2CapClientChannel l2CapChannel = ((L2CapClientChannel) l2CapInfo.getL2CapChannel(device));
        l2CapChannel.connectToL2CapChannel(secure, resultCallback);
    }

    public void read(final ReadL2CapChannelRequest request, final Result resultCallback) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            firePlatformNotSupportedError(resultCallback);
            return;
        }
        final String deviceId = request.remoteId;
        final BluetoothDevice device = adapter.getRemoteDevice(deviceId);
        final int psm = request.psm;
        final L2CapChannel channel = findChannel(psm, device);
        if (channel == null) {
            resultCallback.error(ErrorCodes.NO_OPEN_L2CAP_CHANNEL_FOUND, "No open channel found for device " + device.getAddress() + " / psm " + psm, null);
            return;
        }
        channel.read(request, resultCallback);
    }

    public void write(final WriteL2CapChannelRequest request, final Result resultCallback) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            firePlatformNotSupportedError(resultCallback);
            return;
        }
        final String deviceId = request.remoteId;
        final BluetoothDevice device = adapter.getRemoteDevice(deviceId);
        final int psm = request.psm;
        final L2CapChannel channel = findChannel(psm, device);
        if (channel == null) {
            resultCallback.error(ErrorCodes.NO_OPEN_L2CAP_CHANNEL_FOUND, "No open channel found for device " + device.getAddress() + " / psm " + psm, null);
            return;
        }
        channel.write(request, resultCallback);
    }

    public synchronized void closeChannel(final CloseL2CapChannelRequest request, final Result resultCallback) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            firePlatformNotSupportedError(resultCallback);
            return;
        }
        final String deviceId = request.remoteId;
        final BluetoothDevice device = adapter.getRemoteDevice(deviceId);
        final int psm = request.psm;
        final L2CapInfo channelInfo = findInfo(psm);
        if (channelInfo != null) {
            try {
                channelInfo.close(device);
            } catch (IOException e) {
                LogLevel.ERROR.log(e.getMessage(), e);
                resultCallback.error(ErrorCodes.CLOSE_L2CAP_CHANNEL_FAILED, "Can't close channel with psm " + psm, null);
            }
            if (channelInfo.getType() == L2CapInfo.Type.CLIENT) {
                openL2CapChannelInfos.remove(channelInfo);
            }
        } else {
            LogLevel.DEBUG.log("No channel found which is matching device " + device.getAddress() + " / psm " + psm);
        }
        resultCallback.success(null);
    }

    public synchronized void closeServerSocket(CloseL2CapServer options, final Result resultCallback) {
        final int psm = options.psm;
        final L2CapInfo channelInfo = findInfo(psm);
        if (channelInfo != null && channelInfo.getType() == L2CapInfo.Type.SERVER) {
            ((ServerSocketInfo) channelInfo).closeSocket();
            openL2CapChannelInfos.remove(channelInfo);
        } else {
            LogLevel.DEBUG.log("No server socket found with psm " + psm);
        }
        resultCallback.success(null);
    }

    private void firePlatformNotSupportedError(Result resultCallback) {
        resultCallback.error(ErrorCodes.PLATFORM_NOT_SUPPORTED, "The device is running an older Android version. Minimum version is Android Q.", null);
    }

    @Nullable
    private L2CapInfo findInfo(final int psm) {
        for (L2CapInfo channelInfo : openL2CapChannelInfos) {
            if (psm == channelInfo.getPsm()) {
                return channelInfo;
            }
        }
        return null;
    }

    @Nullable
    private L2CapChannel findChannel(final int psm, final BluetoothDevice remoteDevice) {
        final L2CapInfo l2CapInfo = findInfo(psm);
        if (l2CapInfo == null) {
            return null;
        }
        return l2CapInfo.getL2CapChannel(remoteDevice);
    }


    public interface DeviceConnected {
        void deviceConnected(BluetoothDevice remoteDevice, int psm);
    }
}
